Python/pylifecycle.c

```diff
@@ -1491,6 +1491,22 @@ finalize_modules_delete_special(PyThreadState *tstate, int verbose)
     }
 }
 
+static void
+swap_module_dict(PyModuleObject *mod)
+{
+    if (_Py_IsImmortal(mod->md_dict)) {
+        // gh-117783: Immortalizing module dicts can cause some finalizers to
+        // run much later than typical leading to attribute errors due to
+        // partially cleared modules. To avoid this, we copy the module dict
+        // if it was immortalized.
+        PyObject *copy = PyDict_Copy(mod->md_dict);
```

Comment:
Do we need to clear the original dict here too?  Otherwise it seems like the immortal dict will just hang around and keep everything alive anyway?

Reply:
My intention was to keep the fields alive, but maybe that's not the best way to address this.

Basically, `test_concurrent_futures.test_interpreter_shutdown` was failing because a weakref was cleared later than usual, after the module dictionary was partially cleared. I was concerned that the problem might not be limited to that test, but I don't really have evidence for that one way or the other.

I'll get rid of this swapping and make the weakref callback more robust.